src/libos/src/net/socket/unix/stream/server.rs

```diff
@@ -0,0 +1,110 @@
+use super::*;
+use std::collections::btree_map::BTreeMap;
+
+#[derive(Debug)]
+pub struct Servers {
+    file: SgxMutex<BTreeMap<String, Option<Arc<Server>>>>,
+    abstr: SgxMutex<BTreeMap<String, Option<Arc<Server>>>>,
+}
+
+impl Servers {
+    pub fn new() -> Self {
+        Self {
+            file: SgxMutex::new(BTreeMap::new()),
+            abstr: SgxMutex::new(BTreeMap::new()),
+        }
+    }
+
+    pub fn add_binder(&self, addr: &Addr) -> Result<()> {
+        let key = Self::get_key(addr);
+        let mut servers = self.get_servers(addr);
+        if servers.contains_key(&key) {
+            return_errno!(EADDRINUSE, "the addr is already bound");
+        } else {
+            servers.insert(key, None);
+            Ok(())
+        }
+    }
+
+    pub fn add_listener(&self, addr: &Addr) -> Result<()> {
+        let key = Self::get_key(addr);
+        let mut servers = self.get_servers(addr);
+
+        // The server is updated every time listen is called.
+        if let Some(option) = servers.get(&key) {
+            servers.insert(key, Some(Arc::new(Server::new())));
+            Ok(())
+        } else {
+            return_errno!(EINVAL, "the socket is not bound");
+        }
+    }
+
+    pub fn push_incoming(&self, addr: &Addr, sock: Stream) -> Result<()> {
+        self.get_server(addr)
+            .ok_or_else(|| errno!(ECONNREFUSED, "no one's listening on the remote address"))?
+            .push_incoming(sock);
+        Ok(())
+    }
+
+    pub fn pop_incoming(&self, addr: &Addr) -> Result<Stream> {
+        self.get_server(addr)
+            .ok_or_else(|| errno!(EINVAL, "the socket is not listening"))?
+            .pop_incoming()
+            .ok_or_else(|| errno!(EAGAIN, "No connection is incoming"))
+    }
+
+    pub fn get_server(&self, addr: &Addr) -> Option<Arc<Server>> {
+        let key = Self::get_key(addr);
+        let servers = self.get_servers(addr);
+        servers.get(&key).map(|x| x.clone()).flatten()
+    }
+
+    pub fn remove_server(&self, addr: &Addr) {
+        let key = Self::get_key(addr);
+        let mut servers = self.get_servers(addr);
+        servers.remove(&key);
+    }
+
+    fn get_servers(&self, addr: &Addr) -> SgxMutexGuard<'_, BTreeMap<String, Option<Arc<Server>>>> {
+        match addr {
+            Addr::File(unix_path) => self.file.lock().unwrap(),
+            Addr::Abstract(path) => self.abstr.lock().unwrap(),
+            Addr::Unnamed => todo!("impl autobind for uds"),
+        }
+    }
+
+    fn get_key(addr: &Addr) -> String {
+        match addr {
+            Addr::File(unix_path) => unix_path.absolute(),
+            Addr::Abstract(path) => addr.path_str().to_string(),
+            Addr::Unnamed => unreachable!(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Server {
+    incoming: SgxMutex<VecDeque<Stream>>,
```

Comment:
Suggestion 1. Use `Channel<Stream>` rather than `SgxMutex<VecDeque<Stream>>` so that a blocking accept and the support of backlog for listen can be implemented easily.

Suggestion 2. Logically, the incoming connections of a listening socket belongs to the socket. So I think it makes more sense to maintain the incoming connections in struct `Stream`, not a `Server` struct.

By the way, to avoid confusion, we should not introduce a type named `Server`. Technically, there are no servers here. A server should include logic to handle requests; a listening or passive socket is not a server. Since we are implementing sockets, we should use socket terms, in which "severs" are called listening or passive sockets.

 I think the  should be maintained as part of the `Stream` type, which